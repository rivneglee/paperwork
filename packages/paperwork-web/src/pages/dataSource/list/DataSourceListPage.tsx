import React from 'react';
import { Card, List, Badge } from '@paperwork/ui-widgets';

const DataSourceListPage = () => (
  <div>
    <Card header={<h3>My Datasource</h3>}>
      <List>
        <List.Item>Datasource 1 <Badge color="secondary">Shared</Badge></List.Item>
        <List.Item>Datasource 2</List.Item>
      </List>
    </Card>
    <Card header={<h3>Collaborative Datasource</h3>}>
      <List>
        <List.Item>Datasource 3</List.Item>
        <List.Item>Datasource 4</List.Item>
      </List>
    </Card>
  </div>
);

export default DataSourceListPage;
