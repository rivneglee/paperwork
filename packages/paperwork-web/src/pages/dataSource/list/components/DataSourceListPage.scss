.pwapp-datasource-list {
  &__row {
    flex-grow: 1;
    display: flex;
    justify-content: space-between;
    color: inherit;
    text-decoration: none;
    align-items: center;
  }
}
