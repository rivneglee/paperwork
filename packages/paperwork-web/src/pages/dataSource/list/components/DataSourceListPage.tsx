import React, { FunctionComponent } from 'react';
import { Link } from 'react-router-dom';

import { List, BaseTemplate, Card, Badge, Search, Icons, QuickAdd, Tooltip, PageState } from '@paperwork/ui-widgets';
import AppBar from '../../../../components/AppBar';
import Spinner from '../../../../components/PageTransitionSpinner/Spinner';
import { UserIdentifier } from '../../../../schema/User';
import StickySideBar from '../../../../components/StickySideBar/StickySideBar';
import './DataSourceListPage.scss';
import noResultFoundImg from '../../../../assets/no-results-found.svg';

interface BadgeProps {
  color: 'secondary' | 'primary';
  text: string;
}

interface Entry {
  id: string;
  name: string;
  owner: UserIdentifier;
  badges: BadgeProps[];
  link: string;
}

export interface FilterOption {
  key: string;
  value: number | string;
}

export interface FilterOptions {
  keyword?: string;
}

interface Props {
  entries: Entry[];
  filterOptions: FilterOptions;
  onFilterChange: (option: FilterOption) => void;
  onApplyFilter: (filterOptions: FilterOptions) => void;
  onCreateNew: () => void;
  isProcessing?: boolean;
}

const handleFilterChange = (key: string, handler: any) => (e: any) => handler({ key, value: e.target.value });

const DataSourceListPage: FunctionComponent<Props> = ({
  entries = [],
  isProcessing,
  filterOptions = {},
  onFilterChange,
  onApplyFilter,
  onCreateNew,
}) => (
  <BaseTemplate
    header={<AppBar activeMenuId="datasource" />}
    isProcessing={isProcessing}
    spinner={<Spinner />}
  >
    <Card header={
      <Card.Header primary="My datasource"/>
    }>
      <StickySideBar>
        <Search
          value={filterOptions.keyword}
          placeholder="Search"
          onChange={handleFilterChange('keyword', onFilterChange)}
          onApply={() => onApplyFilter(filterOptions)}
        />
        <QuickAdd color="secondary">
          <QuickAdd.Item onClick={onCreateNew} icon={<Icons.DataSource />} tooltip="Create empty datasource"/>
        </QuickAdd>
      </StickySideBar>
      {
        entries.length > 0 ? (
          <List>
            {
              entries.map(({ name, id, badges, link }) => (
                <List.Item key={id}>
                  <Link to={link} className="pwapp-datasource-list__row">
                    {name}
                    {
                      badges.map(({ color, text }) =>
                        <Tooltip
                          key={text}
                          placement="top"
                          content="You are not the owner of this datasource, but you have limited access to this datasource">
                          <Badge color={color}>{text}</Badge>
                        </Tooltip>,
                      )
                    }
                  </Link>
                </List.Item>
              ))
            }
          </List>
        ) : (
          <PageState
            image={noResultFoundImg}
            title="No results found"
            description="Please try to improve your filter and create some data"
          />
        )
      }
    </Card>
  </BaseTemplate>
);

export default DataSourceListPage;
