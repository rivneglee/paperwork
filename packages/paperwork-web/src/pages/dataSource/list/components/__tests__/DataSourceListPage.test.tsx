import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';

import DataSourceListPage from '../DataSourceListPage';

describe('DataSourceListPage', () => {
  const filterOptions = {};
  const onApplyFilter = jest.fn();
  const onFilterChange = jest.fn();
  const onCreateNew = jest.fn();
  it('should render when result is empty', () => {
    const wrapper: ShallowWrapper = shallow(
      <DataSourceListPage
        onCreateNew={onCreateNew}
        onApplyFilter={onApplyFilter}
        onFilterChange={onFilterChange}
        entries={[]} filterOptions={filterOptions}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should render when result is not empty', () => {
    const wrapper: ShallowWrapper = shallow(
      <DataSourceListPage
        onCreateNew={onCreateNew}
        onApplyFilter={onApplyFilter}
        onFilterChange={onFilterChange}
        entries={[
          {
            id: '1',
            name: 'DS_1',
            owner: '2',
            badges: [{
              color: 'secondary',
              text: 'Collaborative',
            }],
            link: '/1/dataSource/1',
          },
          {
            id: '2',
            name: 'DS_2',
            owner: '1',
            badges: [],
            link: '/1/dataSource/2',
          },
        ]} filterOptions={filterOptions}
      />,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
