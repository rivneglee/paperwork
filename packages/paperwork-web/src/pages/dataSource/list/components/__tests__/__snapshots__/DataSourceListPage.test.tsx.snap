// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataSourceListPage should render when result is empty 1`] = `
<ei
  header={<AppBar />}
>
  <i
    header={
      <h3>
        My Datasource
      </h3>
    }
  >
    <StickySideBar>
      <Qo
        onApply={[Function]}
        onChange={[Function]}
        placeholder="Search"
      />
      <t
        color="secondary"
      >
        <Component
          icon={<DataSource />}
          tooltip="Create empty datasource"
        />
      </t>
    </StickySideBar>
    <ti
      description="Please try to improve your filter and create some data"
      image="no-results-found.svg"
      title="No results found"
    />
  </i>
</ei>
`;

exports[`DataSourceListPage should render when result is not empty 1`] = `
<ei
  header={<AppBar />}
>
  <i
    header={
      <h3>
        My Datasource
      </h3>
    }
  >
    <StickySideBar>
      <Qo
        onApply={[Function]}
        onChange={[Function]}
        placeholder="Search"
      />
      <t
        color="secondary"
      >
        <Component
          icon={<DataSource />}
          tooltip="Create empty datasource"
        />
      </t>
    </StickySideBar>
    <s>
      <Component
        key="1"
      >
        <Link
          className="pwapp-datasource-list__row"
          to="/1/dataSource/1"
        >
          DS_1
          <Component
            content="You are not the owner of this datasource, but you have limited access to this datasource"
            key="Collaborative"
            placement="top"
          >
            <M
              color="secondary"
            >
              Collaborative
            </M>
          </Component>
        </Link>
      </Component>
      <Component
        key="2"
      >
        <Link
          className="pwapp-datasource-list__row"
          to="/1/dataSource/2"
        >
          DS_2
        </Link>
      </Component>
    </s>
  </i>
</ei>
`;
